src/core/createConsent.js

```diff
@@ -0,0 +1,273 @@
+    },
+    refreshFromCookies() {
+      refreshFromCookies();
+      onChange(getStatus());
+    },
+    getStatus
+  };
+};
+
+export default ({
+  config,
+  logger,
+  lifecycle,
+  createConsentRequestPayload,
+  sendEdgeNetworkRequest
+}) => {
+  const { orgId, optInEnabled } = config;
+
+  const deferredsAwaitingConsent = [];
+
+  const processConsentStatus = consentStatus => {
+    if (consentStatus.consented === undefined) {
+      return;
+    }
+
+    while (deferredsAwaitingConsent.length) {
+      const deferred = deferredsAwaitingConsent.shift();
+
+      if (consentStatus.consented) {
+        deferred.resolve();
+      } else {
+        deferred.reject(new Error(consentStatus.deniedConsentReason));
+      }
+    }
+  };
+
+  const consentState = createConsentState({
+    optInEnabled,
+    optInCookieName: getNamespacedCookieName(orgId, OPTIN_COOKIE_KEY),
+    optOutCookieName: getNamespacedCookieName(orgId, OPTOUT_COOKIE_KEY),
+    onChange: processConsentStatus
+  });
+
+  const initialConsentStatus = consentState.getStatus();
+
+  if (initialConsentStatus.consented === undefined) {
+    logger.warn("Some commands may be delayed until the user opts in.");
+  } else if (!initialConsentStatus.consented) {
+    logger.warn(
+      `Some commands may be fail. ${initialConsentStatus.deniedConsentReason}`
+    );
+  }
+
+  return {
+    /**
+     * Update the purposes the user has opted into. Only to be called by the
+     * Privacy component.
+     * @param {Object} preferencesByPurpose An object where the key is
+     * the purposes name and the value is a boolean indicating whether the user
+     * has opted into the purpose.
+     */
+    setOptInPurposes(preferencesByPurpose) {
+      const status = consentState.getStatus();
+
+      // The edge restricts modifying opt-in preferences if the user is opted
+      // out, so we'll prevent the attempt.
+      if (
+        !status.consented &&
+        status.deniedConsentReason === OPTED_OUT_OF_ALL_PURPOSES
+      ) {
+        return Promise.reject(new Error(OPTED_OUT_OF_ALL_PURPOSES));
+      }
+
+      consentState.setOptInPreferences(preferencesByPurpose);
+
+      const payload = createConsentRequestPayload();
+      return lifecycle
+        .onBeforeConsentRequest({
+          payload
+        })
+        .then(() => {
+          payload.setPurposes(preferencesByPurpose);
+          return sendEdgeNetworkRequest({
+            payload,
+            action: "opt-in"
+          });
+        })
+        .then(() => {
+          // We optimistically use the preferences the user provided right away,
+          // but if the server doesn't agree and sends back a cookie value that
+          // conflicts, we adopt whatever the server said since that ultimately
+          // is what will be used if the user refreshes the page.
+          consentState.refreshFromCookies();
+        })
+        .catch(error => {
+          consentState.refreshFromCookies();
```

Comment:
If the consent endpoint is not working correctly, this stops us from collecting data.  This especially becomes a problem in a SPA.  Why do you call refreshFromCookies after the response?  Why would the server set the consent cookie to anything different?